src/main/java/tictactoe/Board.java

```diff
@@ -97,9 +96,14 @@ private int getSide() {
         return (int)Math.sqrt(grid.size());
     }
 
-    public void markCellAtPosition(String mark, int position) {
-        Cell cell = grid.get(position - 1);
-        cell.setMark(mark);
+    public void markCellAtPosition(String mark, String input) {
+        if(isValidInput(input)) {
+            int position = Integer.parseInt(input);
+            Cell cell = grid.get(position - 1);
+            cell.setMark(mark);
+        } else {
+            throw new InvalidInputException("Invalid input, please enter a number between 1 and 9.");
```

Comment:
This is another instance of the Board being aware of the Input/Output details.

The exception could instead be an object that contains a `reason` Enum and perhaps a `validInput` List. It would then be up to whichever class is using the Board to present the error via whatever Input/Output is in use.

The same goes for the Cell class.

Also, you always ask to enter a number between 1 and 9, but if say 3 is already taken, then choosing 3 will be frustrating in terms of user experience, as you just told me that anything between 1 and 9 is fine.